include/oneapi/tbb/collaborative_call_once.h

```diff
@@ -0,0 +1,185 @@
+namespace detail {
+namespace d1 {
+
+#if _MSC_VER && !defined(__INTEL_COMPILER)
+    // Suppress warning: structure was padded due to alignment specifier
+    #pragma warning (push)
+    #pragma warning (disable: 4324)
+#endif
+
+constexpr std::size_t bit_count = 7;
+
+static_assert(1 << bit_count == max_nfs_size, "bit_count must be a log2(max_nfs_size)");
+
+inline constexpr std::uintptr_t maskoff_pointer(std::uintptr_t ptr) {
+    return ptr & (std::size_t(-1) << bit_count);
+}
+
+class alignas(max_nfs_size) run_once {
+    task_arena my_arena;
+    wait_context my_wait_context;
+    task_group_context my_context;
+
+    std::atomic<std::int64_t> ref_count{0};
+
+    template <typename Func>
+    class call_once_delegate : public delegate_base {
+    public:
+        call_once_delegate(Func& f) : my_func(f) {}
+
+        bool operator()() const override {
+            my_func();
+            return true;
+        }
+
+    private:
+        Func& my_func;
+    };
+
+    template<typename Fn>
+    void isolated_execute(Fn f) {
+        call_once_delegate<Fn> delegate(f);
+
+        r1::isolate_within_arena(delegate, reinterpret_cast<std::intptr_t>(this));
+    }
+
+public:
+    run_once()
+        : my_arena(task_arena::attach{})
+        , my_wait_context(0)
+        , my_context(task_group_context::bound,
+            task_group_context::default_traits | task_group_context::concurrent_wait)
+        {}
+
+    ~run_once() {
+        spin_wait_while(ref_count, [&](std::int64_t value) { return value > 0; }, std::memory_order_acquire);
+    }
+
+    void increase_ref() { ref_count++; }
+
+    void decrease_ref() { ref_count--; }
+
+    template <typename F>
+    void run_and_wait(F&& f) {
+        my_arena.execute([&] {
+            isolated_execute([&] {
+                function_stack_task<F> t{ std::forward<F>(f), my_wait_context };
+                my_wait_context.reserve();
+                
+                execute_and_wait(t, my_context, my_wait_context, my_context);
+            });
+        });
+    }
+
+    void wait() {
+        my_arena.execute([&] {
+            isolated_execute([&] {
+                // We do not want to get an exception from user functor on moonlighting threads.
+                // The exception is handled with the winner thread
+                task_group_context stub_context;
+                detail::d1::wait(my_wait_context, stub_context);
+            });
+        });
+    }
+
+};
+
+class collaborative_once_flag : no_copy {
+    enum state {uninitialized, done};
+    std::atomic<std::uintptr_t> my_state{ state::uninitialized };
+
+    template <typename Fn, typename... Args>
+    friend void collaborative_call_once(collaborative_once_flag& flag, Fn&& f, Args&&... args);
+
+    template <typename Fn>
+    void do_collaborative_call_once(Fn&& f) {
+        std::uintptr_t expected = my_state.load(std::memory_order_acquire);
+        run_once local_runner;
+        while (expected != state::done) {
```

Comment:
@alexey-katranov does it makes sense to make an early exit before initialization of runner?
As for the second point: it is necessary to create a runner on stack of each thread, they are competing to publish address of their `local_runner`. Thus, shared runner will be published already on the `compare_exchange_strong`.

Reply:
Right. I see several ways we could go:
1. Leave it as is. (Seems like it leaves an easy improvement on the table.)
2. Add my `if`, allowing subsequent calls to return ASAP when the flag is "used".
3. We could additionally have the definitely-moonlighting threads not construct theirs. That is, if on the first pass through, don't construct `local_runner` if we immediately see that `expected` is already pointing at something. That's a bit more work, that maybe could worth if there wind up being lots of moonlighers. But probably not since worst case it's only the number of worker threads.
4. Try to avoid unnecessary construction in the race case: That would basically require another bit and have the "winner" set the bit to "I'm busy constructing the `local_runner`" which others would have to spin-wait on. That actually might be simpler code? Not sure if it's worth it, though.

My preference is for 2 unless you guys think 3 or 4 is worth the complexity.

Reply:
@BenFrantzDale, good catch. I think 4 can be easily implemented with the following wrapper:
```c++
template <T>
class thread_safe_optional {
    std::atomic<bool> m_initialized{false};
    union {
        T m_value;
    }

    thread_safe_optional( const atomic_optional & other ) = delete;
public:
    thread_safe_optional() = default;
    ~thread_safe_optional() {
        if (m_initialized.load(std::memory_order_relaxed)) {
            m_value.~T();
        }
    }
    template <class... Args>
    T& unsafe_emplace( Args&&... args ) {
        new (&m_value) T(std::forward<Args>(args)...);
        return m_value;
    }
    T& get() {
        spin_wait_while_eq(m_initialized, false);
        return m_value;
    }
};

Reply:
I like that. That lets us avoid duplicate constructions and also streamline the already-called case.

Are you missing an `m_initialized.store(std::memory_order_release)` after `new`?

I don't love the name `unsafe_emplace` since it's not obvious what's unsafe about it, especially since it's a `thread_safe_optional`. Maybe `forcibly_emplace` (to emphasize that it's potential UB if it's already emplaced)?

I'd put a comment on `get()` that it's `get()` in the `std::future` sense (blocking). I with no context, I expect `get()` to return a pointer (ala the shared pointers). Maybe clearer from the calling side would be to split it apart and have
```
//! wait() like std::future:
void wait() const { spin_wait_while_eq(m_initialized, false); }
//! value() like std::optional:
T& value() { return m_value; }
```
Then the winning thread would `opt.force_emplace();` and the loosing thread would `opt.wait(); auto& foo = opt.value();`?


Is the `std::memory_order_relaxed` (rather than `std::memory_order_acquire`) in the d'tor because the assumption is that the destroying thread that matters is the one that emplaced?

Reply:
> Are you missing an m_initialized.store(std::memory_order_release) after new?

Yes, thank you!

> I don't love the name unsafe_emplace since it's not obvious what's unsafe about it,

Typically, we use `unsafe_` prefix for not thread safe methods in concurrent containers. In any case, it is just a draft and definitely requires better naming.

> Then the winning thread would opt.force_emplace(); and the loosing thread would opt.wait(); auto& foo = opt.value();?

It seems reasonable

> Is the std::memory_order_relaxed (rather than std::memory_order_acquire) in the d'tor because the assumption is that the destroying thread that matters is the one that emplaced?

Yes, moreover, we expect that the pointer to `thread_safe_optional` is distributed across the threads with happens-before semantics, i.e. even dtor on another thread should be Ok with relaxed.

Overall, it can be better not to introduce an isolated wrapper because it will require careful semantics polishing. The similar methods can be introduced for `run_once` with narrow logic required for the particular case.

Reply:
I tried to introduce initialization logic for runner class directly. Now `task_arena` and contexts can only be created by winner thread by calling `make_runner()` method.
Moonlighters must wait in method `wait_for_init()` for initialization of storage with `task_arena`.